### ISSUE TITLE
Same options and data, different result with different computers.

### ISSUE THREAD

**Issue by kritollm** [cazala/synaptic#138]
I know the result will be different because of random numbers, but there was always more iterations on the fastest computer. Is there a way to ensure same result with equal options and training data?

Windows 10 Intel Core i7 @ 2.60GHz:

iterations 1000 error 0.5832954855597182 rate 0.035
iterations 2000 error 0.47984523349249425 rate 0.035
iterations 3000 error 0.4096766632625817 rate 0.035
iterations 4000 error 0.38598258320892237 rate 0.035
iterations 5000 error 0.2839244462200957 rate 0.035
iterations 6000 error 0.21731707647510384 rate 0.035
iterations 7000 error 0.14974001959676705 rate 0.035
iterations 8000 error 0.0922783295181048 rate 0.035
iterations 9000 error 0.07411138184981307 rate 0.035
iterations 10000 error 0.06626638341129625 rate 0.035
iterations 11000 error 0.06183108647527365 rate 0.035
Win Win
Loss Win
Win Win
Win Win
Win Loss
Win Loss
Win Win
Loss Win
Win Win
Win Win
Loss Loss
Loss Loss
Win Loss
Win Loss
Win Win
Loss Loss

OS X El Capitan 1.8 GHz Intel core i7:

iterations 1000 error 0.5473525119076491 rate 0.035
iterations 2000 error 0.4808354441834457 rate 0.035
iterations 3000 error 0.4521193897218217 rate 0.035
iterations 4000 error 0.43399835084618216 rate 0.035
iterations 5000 error 0.4050580287584716 rate 0.035
iterations 6000 error 0.29390403125518283 rate 0.035
iterations 7000 error 0.16752626336749996 rate 0.035
iterations 8000 error 0.0810127505036039 rate 0.035
iterations 9000 error 0.06531469909494222 rate 0.035
Loss Win
Win Win
Loss Win
Loss Win
Win Loss
Loss Loss
Loss Win
Loss Win
Loss Win
Win Win
Loss Loss
Loss Loss
Win Loss
Win Loss
Loss Win
Loss Loss

**Comment by Jabher** [comment]
Can you share the code you're using please?

**Comment by kritollm** [comment]
Yes, but do you need the training data?

**Comment by Jabher** [comment]
Let's look at code first

**Comment by kritollm** [comment]
```` javascript
const synaptic = require('synaptic');
let brain = null
function train(a) {
    let trainData = a.map(e => ({ input: e.hh.concat(e.ho, e.ah, e.ao), output: [e.w] }));
    //const alpha = 3;
    //let hidden = Math.round(trainData.length / (alpha * (48 + 1)));
    //let hidden = Math.round(((2 * 48) / (3 * 3)) + 1);
    //console.log('hidden: ' + hidden);
    brain = new synaptic.Architect.Perceptron(30, 24, 1);
    let trainer = new synaptic.Trainer(brain);
    trainer.train(trainData, {
        rate: 0.035,
        iterations: 30000,
        error: 0.06,
        log: 1000,
        cost: synaptic.Trainer.cost.CROSS_ENTROPY
    });
}
function predict(market) {
    market.forEach(e => {
        //let data = e.ho.concat(e.ao);
        let data = e.hh.concat(e.ho, e.ah, e.ao);
        let key = brain.activate(data)[0];

        console.log(key > 0.5 ? "Win" : "Loss", e.w ? "Win" : "Loss");

    });
}

module.exports = exports = { train, predict };

**Comment by Jabher** [comment]
What are node.js versions on a macos and win computers?

**Comment by kritollm** [comment]
6.5 on both.

**Comment by Jabher** [comment]
I'd rather suggest that it is somehow related to OS. 
Number of iterations should be same to achieve the lower values, thing that should differ (for same code) is only the time of execution. 
From an FP point of view, we can assume that this is a pure function that always returns same values for same inputs, if we do assume that we have 2 inputs - train data and **random generator output**.

On Windows, V8 (and node.js) uses rand_s function that is Windows-specific way of providing cryptographically-secure random numbers (http://msdn.microsoft.com/library/sxtz2fa8(v=vs.90).aspx).

**Comment by kritollm** [comment]
If none of the calculations depend on time used, we can assume it's the os spesifik random generator. The random generator probably use the clock, but it's strange that all the training sets used more iterations to reach the minimum error rate on the fastest computer. It was faster in time, but used more iterations to complete. I can test it on Windows 10 on my Mac later.

**Comment by Jabher** [comment]
If you'll wait long enough you'll see how random number in this project will become configurable with one of the options available with the seed (for reproducability) :)

**Comment by Jabher** [comment]
BTW, I think that's what can explain everything:
https://github.com/nodejs/node/blob/master/src/node_crypto.cc#L264-L266

**Comment by kritollm** [comment]
Yes. I think that explain it :+1: 

So synaptic could use another rand function when the code run in node?

It's really okay, you just make sure to save the best versions. I like that I can explore ML with JS. You have done advanced technology easy to use. I also get more control using my favorite language than hacking some R scripts.

**Comment by Jabher** [comment]
Theoretically - yes, it is not a big deal.
Great, I'm having a same reason to watch over this project
